cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+                              num_batches,
+                              bloom_filter_->partitions[partition_idx_].get()));
+                
+            return callbacks_.start_task_group(
+                task_group_bloom_[partition_idx_],
+                build_accumulator_.batch_count(partition_idx_));
+        }
+        
+        Status SpillingHashJoin::OnBloomFilterFinished(size_t thread_index)
+        {
+            RETURN_NOT_OK(callbacks_.bloom_filter_finished(thread_index));
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
+                return StartCollocatedJoins(thread_index);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnBloomFiltersReceived(size_t thread_index)
+        {
+            bloom_ready_.store(true, std::memory_order_release);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnProbeSideBatch(size_t thread_index, ExecBatch batch)
+        {
+            if(bloom_ready_.load())
+            {
+                RETURN_NOT_OK(callbacks_.apply_bloom_filter(
+                    thread_index,
+                    &batch));
+            }
+            return probe_accumulator_.InsertBatch(
+                              thread_index,
+                              std::move(batch));
+        }
+
+        Status SpillingHashJoin::OnProbeSideFinished(size_t thread_index)
+        {
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
+                return StartCollocatedJoins(thread_index);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::StartCollocatedJoins(size_t thread_index)
+        {
+            partition_idx_ = SpillingAccumulationQueue::kNumPartitions;
+            return BeginNextCollocatedJoin(thread_index);
+        }
+
+        Status SpillingHashJoin::BeginNextCollocatedJoin(size_t thread_index)
+        {
+            partition_idx_ -= 1;
+            probe_or_hash_table_ready_.store(false);
+            return build_accumulator_
+                .GetPartition(
+                    thread_index,
+                    partition_idx_,
+                    [this](size_t thread_index, AccumulationQueue queue)
+                    {
+                        return BuildHashTable(thread_index, std::move(queue));
+                    });
+        }
+
+        Status SpillingHashJoin::BuildHashTable(size_t thread_index, AccumulationQueue queue)
+        {
+            RETURN_NOT_OK(
+                impls_[partition_idx_]->BuildHashTable(
+                    thread_index,
+                    std::move(queue),
+                    [this](size_t thread_index)
+                    {
+                        return OnHashTableFinished(thread_index);
+                    }));
+            RETURN_NOT_OK(
+                probe_accumulator_
+                .GetPartition(
```

Comment:
Are we reading the entire probe side partition into memory before starting lookups? Probe side should be streaming, since it may be a lot larger than build side.